.github/workflows/image-builder.yml

```diff
@@ -0,0 +1,181 @@
+      folder:
+        description: 'Either "workspaces" or "graders"'
+        required: true
+        type: string
+      image-prefix:
+        description: 'The prefix for the image name (e.g. prairielearn/workspace)'
+        required: true
+        type: string
+      base-images:
+        description: 'The base images to build'
+        required: false
+        type: string
+        default: '[]'
+      images:
+        description: 'The images to build'
+        required: true
+        type: string
+      exclude:
+        description: 'The combinations to exclude'
+        required: false
+        type: string
+        default: '[]'
+
+jobs:
+  build-base-images:
+    strategy:
+      max-parallel: 5
+      matrix:
+        image: ${{ fromJSON(inputs.base-images) }}
+        os: [ubuntu-24.04, ubuntu-24.04-arm]
+        exclude: ${{ fromJSON(inputs.exclude) }}
+    runs-on: ${{ matrix.os }}
+    env:
+      CAN_PUSH_IMAGE: ${{ secrets.DOCKERHUB_USERNAME != '' && github.ref_name == github.event.repository.default_branch }}
+    steps:
+      # Checkout the repository with full depth so we can check if the workspace was modified
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check whether ${{ inputs.folder }}/${{ matrix.image }} was modified
+        id: check-modified
+        run: |
+          scripts/check_path_modified.sh ${{ inputs.folder }}/${{ matrix.image }} ${{ inputs.folder }}_${{ matrix.image }}_modified
+
+      - name: Set up Docker Buildx
+        # Currently, we need this 'if' on each step. https://github.com/orgs/community/discussions/82744
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to DockerHub
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' && env.CAN_PUSH_IMAGE == 'true' }}
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Set docker tags and labels
+        id: meta
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/metadata-action@v5
+        with:
+          context: git
+          images: |
+            ${{ inputs.image-prefix }}-${{ matrix.image }}
+          # On the default branch, we'll also tag the image as `latest`
+          tags: |
+            type=raw,value=latest,enable={{is_default_branch}}
+            type=sha,prefix=,format=long
+
+      - name: Build ${{ (env.CAN_PUSH_IMAGE == 'true' && 'and push') || '' }}
+        id: build
+        if: ${{ steps.check-modified.outputs.was-modified == 'true' }}
+        uses: docker/build-push-action@v6
+        with:
+          context: ${{ inputs.folder }}/${{ matrix.image }}
+          labels: ${{ steps.meta.outputs.labels }}
+          tags: ${{ steps.meta.outputs.tags }}
+          push: ${{ env.CAN_PUSH_IMAGE == 'true' }}
```

Comment:
I don't believe this setup handles multi-platform images correctly. This will have last-one-wins behavior, where whichever platform (x86 or ARM) that finishes building last will end up being the only image available under the `latest` tag.

To do this correctly, I believe you need to use the approach described on https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners, where you use `buildx imagetools create` after the individual platforms have been built and pushed.

Ditto for the non-base-image jobs.

Reply:
Good catch, sorry about that. Should be fixed in [79b8b71](https://github.com/PrairieLearn/PrairieLearn/pull/11102/commits/79b8b7131665a4df20d7929a966b367db7e2c671).